src/core/net/dns_client.cpp

```diff
@@ -815,20 +860,60 @@ void Client::SendQuery(Query &aQuery, QueryInfo &aInfo, bool aUpdateTimer)
         SuccessOrExit(error = message->Append(Question(kQuestionRecordTypes[aInfo.mQueryType][num])));
     }
 
-    messageInfo.SetPeerAddr(aInfo.mConfig.GetServerSockAddr().GetAddress());
-    messageInfo.SetPeerPort(aInfo.mConfig.GetServerSockAddr().GetPort());
+    length = message->GetLength() - message->GetOffset();
+#if OPENTHREAD_CONFIG_DNS_OVER_TCP_ENABLE
 
-    SuccessOrExit(error = mSocket.SendTo(*message, messageInfo));
+    // Send udp query in case of message being greater than kUdpQueryMaxSize or as last resort.
+    if (length >= kUdpQueryMaxSize || aInfo.mTransmissionCount >= aInfo.mConfig.GetMaxTxAttempts() ||
+        aInfo.mConfig.GetTransportProto() == QueryConfig::kDnsTcpTransport)
```

Comment:
I would say that we can configure this in compile time, but this adds not necessary configuration complexity on end developer site. 
So I would propose to drop `aInfo.mTransmissionCount >= aInfo.mConfig.GetMaxTxAttempts()` and leave the rest as it is since it's compliant with RFC.

Reply:
fixed